woocommerce/woocommerce-ios
# Upsell Card Reader Banner in Orders

Closes: #7249


⛔ Do not merge until [e2a89d2381dfad83df0f0a236d4e16384e44a46f](https://github.com/woocommerce/woocommerce-ios/commit/599cd6bb74e6410506cbce5c0c98508e8fea2814) is reverted: This is just to make testing easier. With that test code we:
- Reset the banner policy show when starting the app, so it can be shown again even if the user dismissed it before.
- Set the remind me later to 20 seconds instead of 14 days, so it can be tested faster.
(Reverted)


### Description

This banner is basically the same we implemented and show in the Payment Methods and App Settings screens, with the call to action button that we used on Payments Methods as well, sending the user to the card reader shop. Please notice that the banner show policy is the same as the one shown in the Payment Methods screen, so if it is dismissed (forever or just to be reminded later) there, it won't be shown in the settings or payment methods. We also have the same flow here, when the user presses on [X] we show an alert to choose between dismissing it forever or just being reminded later on.
Before we could already show some banners on top of this order list, either the error banner or the orders survey feature. The logic for the banners is:
- Only one banner can be shown at once.
- In case we want to show more than one banner, the priority goes as: Error, Upsell Card Readers, Orders Survey.
- On landscape we do not show the Upsell Card Readers banner

### Testing instructions


1. Navigate to Orders tab
2. Tap on Purchase Card Reader. The Card Readers shop web view opens.
3. Observe that the Upsell Card Reader banner is shown, with a x button to dismiss.
4. Tap the x button and then the `Remind me later` option – observe that the card is dismissed
5. Go back and forth to orders again within the remind time, and observe that the banner doesn't show there.
6. Go back and forth to orders and open it again after the remind time elapses, observe that the card is shown.
7. Tap on x and Don't show again.

Repeat the above test using `Don't show again`, and observe that the card is not shown even after the remind time elapses.

If you want to test the error and Orders banners:

- To show the error banner, add ` return .error` to this [line](https://github.com/woocommerce/woocommerce-ios/blob/819a4cac60a6d70b74f70c6e2809867a775dea7d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift#L296)
- To show the orders banner replace the completion block with `onCompletion(Result.success(true))` [here](https://github.com/woocommerce/woocommerce-ios/blob/5f52e5f7ea41a42c448783ae4b93e6f440a1d425/Yosemite/Yosemite/Stores/AppSettingsStore.swift#L232
) As stated above, the orders banner should not show if the upsell card reader is presented.
  
### Screenshots


https://user-images.githubusercontent.com/1864060/180456139-e9b7e4aa-2496-420d-8843-a7825a223416.mp4




---
